climada/util/interpolation.py

```diff
@@ -186,3 +202,95 @@ def index_nn_haversine(centroids, coordinates, threshold=THRESHOLD):
+                       check_antimeridian=True):
+    """Compute the neareast centroid for each coordinate using a k-d tree.
+
+    Parameters
+    ----------
+    centroids : 2d array
+        First column contains latitude, second column contains longitude.
+        Each row is a geographic point
+    coordinates : 2d array
+        First column contains latitude, second column contains longitude. Each
+        row is a geographic point
+    threshold : float
+        distance threshold in km over which no neighbor will be found. Those
+        are assigned with a -1 index
+    check_antimedirian: bool
+        if True, recomputes the nn in a strip with lon size equal to threshold
+        around the antimeridian using the Haversine distance.
+
+    Returns
+    -------
+    np.array
+        with so many rows as coordinates containing the centroids indexes
+    """
+    # Construct tree from centroids
+    tree = sp.spatial.KDTree(np.radians(centroids))
+    # Select unique exposures coordinates
+    _, idx, inv = np.unique(coordinates, axis=0, return_index=True,
+                            return_inverse=True)
+
+    # query the k closest points of the n_points using dual tree
+    dist, assigned = tree.query(np.radians(coordinates[idx]), k=1, p=2, workers=-1)
+
+    # Raise a warning if the minimum distance is greater than the
+    # threshold and set an unvalid index -1
+    num_warn = np.sum(dist * EARTH_RADIUS_KM > threshold)
+    if num_warn:
+        LOGGER.warning('Distance to closest centroid is greater than %s'
+                       'km for %s coordinates.', threshold, num_warn)
+        assigned[dist * EARTH_RADIUS_KM > threshold] = -1
+
+    if check_antimeridian:
+        assigned = index_antimeridian_strip(np.radians(centroids),
+                                            np.radians(coordinates[idx]),
+                                            threshold,
+                                            assigned)
+
+    # Copy result to all exposures and return value
+    return np.squeeze(assigned[inv])
+
+def index_antimeridian_strip(centroids, coordinates, threshold, assigned):
+    """Recompute nearest neighbors close to the anti-meridian with the
+    Haversine distance
+
+    Parameters
+    ----------
+    centroids : 2d array
+        First column contains latitude, second column contains longitude.
+        Each row is a geographic point
+    coordinates : 2d array
+        First column contains latitude, second column contains longitude. Each
+        row is a geographic point
+    threshold : float
+        distance threshold in km over which no neighbor will be found. Those
+        are assigned with a -1 index
+    assigned : 1d array
+        coordinates that have assigned so far
+
+    Returns
+    -------
+    np.array
+        with so many rows as coordinates containing the centroids indexes
+
+    """
+
+    coord_strip_bool =  coordinates[:, 1] + np.pi < 1.5*threshold/EARTH_RADIUS_KM
```

Comment:
This assumes that all centroids and coordinates have longitudes in `[-180, 180]`. The recomputation will only fix assignments close to lon=180. If, however, the coordinates are in `[0, 360]` (or something completely different), the recomputation will not make it worse, but it will also fail to fix the assignments close to lon=0.

One solution would be for `index_antimeridian_strip` to automatically guess the `mid_lon` from the input coordinates, by computing:
```
lon_min = min(centroids[:, 1].min(), coordinates[:, 1].min())
lon_max = max(centroids[:, 1].max(), coordinates[:, 1].max())
if lon_max - lon_min > 180:
    raise ValueError("...")
mid_lon = 0.5 * (lon_max - lon_min)
antimeridian = mid_lon + 180
```

Reply:
Good point. It is in CLIMADA sometimes implicitly assumed that the range is `[-180, 180]`, but it mustn't be the case.

Reply:
I think it should be `mid_lon = 0.5 * (lon_max + lon_min)` though.

Reply:
Yes, sure. Well spotted :+1: